### PR TITLE
Remove various synchronisations from cuda API calls, instrument monitor

### PR DESCRIPTION
**Pull request by RAMitchell**
with nvtx profiler ranges.

@canonizer @trivialfis so far we have been relying on the monitor class with synchronisations in order to estimate the performance of different parts of the gpu_hist algorithm. I have come to realise that the synchronisation itself is a major problem, so I want to rely less on this and use the profiler more.

When built with nvtx the ranges covered by monitor will now clearly show up on the nsight profiler.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -10,9 +10,12 @@ msvc_use_static_runtime()
 # Options
 ## GPUs
 option(USE_CUDA  "Build with GPU acceleration" OFF)
+option(INSTRUMENT_CUDA  "Build with cuda profiling annotations. Developers only." OFF)
```
Comment (canonizer) [inline CMakeLists.txt]: Maybe call it `USE_NVTX`, to keep the name similar to other such constants?

File: doc/gpu/index.rst [hunk 2]
```diff
@@ -195,6 +195,10 @@ Training time time on 1,000,000 rows x 50 columns with 500 boosting iterations a
 
 See `GPU Accelerated XGBoost <https://xgboost.ai/2016/12/14/GPU-accelerated-xgboost.html>`_ and `Updates to the XGBoost GPU algorithms <https://xgboost.ai/2018/07/04/gpu-xgboost-update.html>`_ for additional performance benchmarks of the ``gpu_exact`` and ``gpu_hist`` tree methods.
 
+Developer notes
+==========
+The application may be profiled with annotations by specifying INSTRUMENT_CUDA to cmake and providing the path to the stand-alone nvtx header via NVTX_HEADER_DIR. Regions covered by the 'Monitor' class in cuda code will automatically appear in the nsight profiler.
```
Comment (trivialfis) [inline doc/gpu/index.rst]: Update this as well. `INSTRUMENT_CUDA` -> `USE_NVTX`

**Comment by hcho3** [comment]
@RAMitchell @trivialfis I recall seeing #3799 where NCCL cause NVProf to fail. Was the issue solved?

**Comment by RAMitchell** [comment]
I haven't even tried profiling the multi-GPU version yet. I will check it before merging.

**Pull request merged**